internal/controller/replication.storage/volumegroupreplication_controller.go

```diff
@@ -18,36 +18,752 @@ package controller
+					if err != nil {
+						logger.Error(err, "failed to parse selector from VolumeGroupReplication's annotation")
+						return []reconcile.Request{}
+					}
+					objLabelsSet := labels.Set(objLabels)
+					if labelSelector.Matches(objLabelsSet) {
+						return []reconcile.Request{
+							{
+								NamespacedName: types.NamespacedName{
+									Namespace: vgr.Namespace,
+									Name:      vgr.Name,
+								},
+							},
+						}
+					}
+				}
+			}
+
+			return []reconcile.Request{}
+		},
+	)
+
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&replicationv1alpha1.VolumeGroupReplication{}).
+		For(&replicationv1alpha1.VolumeGroupReplication{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Owns(&replicationv1alpha1.VolumeGroupReplicationContent{}, builder.WithPredicates(skipUpdates)).
+		Owns(&replicationv1alpha1.VolumeReplication{}, builder.WithPredicates(skipUpdates)).
+		Watches(&replicationv1alpha1.VolumeGroupReplicationContent{}, enqueueVGRRequest, builder.WithPredicates(watchOnlySpecUpdates)).
+		Watches(&replicationv1alpha1.VolumeReplication{}, enqueueVGRRequest, builder.WithPredicates(watchOnlyStatusUpdates)).
+		Watches(&corev1.PersistentVolumeClaim{}, enqueueVGRForPVCRequest, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
 		Complete(r)
 }
+
+// waitForGroupCrds waits for dependent CRDs to the available in the cluster
+func (r *VolumeGroupReplicationReconciler) waitForGroupCrds() error {
+	err := WaitForVolumeReplicationResource(r.Client, r.log, volumeGroupReplicationClass)
+	if err != nil {
+		r.log.Error(err, "failed to wait for VolumeGroupReplicationClass CRD")
+		return err
+	}
+
+	err = WaitForVolumeReplicationResource(r.Client, r.log, volumeGroupReplication)
+	if err != nil {
+		r.log.Error(err, "failed to wait for VolumeGroupReplication CRD")
+		return err
+	}
+
+	err = WaitForVolumeReplicationResource(r.Client, r.log, volumeGroupReplicationContent)
+	if err != nil {
+		r.log.Error(err, "failed to wait for VolumeGroupReplicationContent CRD")
+		return err
+	}
+
+	return nil
+}
+
+// setGroupReplicationFailure sets the failure replication status on the VolumeGroupReplication resource
+func (r *VolumeGroupReplicationReconciler) setGroupReplicationFailure(
+	instance *replicationv1alpha1.VolumeGroupReplication, err error) error {
+
+	instance.Status.State = GetCurrentReplicationState(instance.Status.State)
+	instance.Status.Message = replication.GetMessageFromError(err)
+	instance.Status.ObservedGeneration = instance.Generation
+	if err := r.Client.Status().Update(r.ctx, instance); err != nil {
+		r.log.Error(err, "failed to update volumeGroupReplication status", "VGRName", instance.Name)
+		return err
+	}
+
+	return nil
+}
+
+// getMatchingPVCsFromSource fecthes the PVCs based on the selectors defined in the VolumeGroupReplication resource
+func (r *VolumeGroupReplicationReconciler) getMatchingPVCsFromSource(instance *replicationv1alpha1.VolumeGroupReplication,
```

Comment:
Why return the pvc-handles too? It feels that this function does more than it needs to do. It is a little cleaner to have a function to fetch the handle from a PVC and call that when the pvc-handle is needed.

Reply:
Done, created a separate function for pvHandles